src/Features/LanguageServer/Protocol/Handler/Diagnostics/DocumentPullDiagnosticHandler.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics
+{
+    [ExportLspMethod(MSLSPMethods.DocumentPullDiagnosticName, mutatesSolutionState: false), Shared]
+    internal class DocumentPullDiagnosticHandler : AbstractPullDiagnosticHandler<DocumentDiagnosticsParams, DiagnosticReport>
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public DocumentPullDiagnosticHandler(
+            ILspSolutionProvider solutionProvider,
+            IDiagnosticService diagnosticService)
+            : base(solutionProvider, diagnosticService)
+        {
+        }
+
+        public override TextDocumentIdentifier? GetTextDocumentIdentifier(DocumentDiagnosticsParams diagnosticsParams)
+            => diagnosticsParams.TextDocument;
+
+        protected override DiagnosticReport CreateReport(TextDocumentIdentifier? identifier, VSDiagnostic[]? diagnostics, string? resultId)
+            => new DiagnosticReport { Diagnostics = diagnostics, ResultId = resultId };
+
+        protected override ImmutableArray<Document> GetOrderedDocuments(RequestContext context)
+        {
+            // For the single document case, that is the one to process.
+            return context.Document == null ? ImmutableArray<Document>.Empty : ImmutableArray.Create(context.Document);
```

Comment:
It is surprising to me that document could be null here.

Reply:
i will doc.  this can def happen.  The core reason for this is that  a client can call into us for a document that we have since removed from the workspace.  They don't know that it's gone, and we need to respond back with "we don't even know about this file anymore".